src/heap.cc

```diff
@@ -333,87 +327,72 @@ void ObjectHeap::install_heap_limit() {
 }
 
 void ObjectHeap::process_registered_vm_finalizers(RootCallback* ss, LivenessOracle* from_space) {
-  process_registered_finalizers_helper(&registered_vm_finalizers_, ss, from_space, this);
-}
-
-void ObjectHeap::process_registered_finalizers_helper(FinalizerNodeFifo* list, RootCallback* ss, LivenessOracle* from_space, ObjectHeap* heap) {
-  // Process the registered finalizer list.
-  CAPTURE3(
-      RootCallback*, ss,
-      LivenessOracle*, from_space,
-      ObjectHeap*, heap);
-  list->remove_wherever([capture](FinalizerNode* node) -> bool {
-    bool is_alive = node->alive(capture.from_space);
-    if (is_alive) {
-      node->roots_do(capture.ss);
-      return false;  // Keep node in list.
-    }
-    node->handle_not_alive(capture.ss, capture.heap);
-    return true; // Remove node from list.
+  process_registered_finalizers_helper(&registered_vm_finalizers_, ss, from_space, false);
+}
+
+void ObjectHeap::process_finalizer_queue(RootCallback* ss, LivenessOracle* from_space) {
+  process_registered_finalizers_helper(&runnable_finalizers_, ss, from_space, true);
+}
+
+void ObjectHeap::process_registered_finalizers_helper(FinalizerNodeFifo* list, RootCallback* ss, LivenessOracle* from_space, bool in_closure_queue) {
+  list->remove_wherever([ss, from_space, in_closure_queue](FinalizerNode* node) -> bool {
+    return node->weak_processing(in_closure_queue, ss, from_space);
   });
 }
 
-bool ObjectHeap::has_finalizer(HeapObject* key, Object* lambda) {
-  for (FinalizerNode* node : registered_toit_finalizers_) {
-    if (node->has_key(key)) return true;
-  }
-  for (FinalizerNode* node : registered_vm_finalizers_) {
-    if (node->has_key(key)) return true;
-  }
-  return false;
+void ObjectHeap::iterate_finalization_roots(RootCallback* cb) {
+  for (auto finalizer : registered_callback_finalizers_) finalizer->roots_do(cb);
+  for (auto finalizer : registered_vm_finalizers_) finalizer->roots_do(cb);
+  for (auto finalizer : runnable_finalizers_) finalizer->roots_do(cb);
 }
 
-bool ObjectHeap::add_toit_finalizer(HeapObject* key, Object* lambda) {
+bool ObjectHeap::add_callable_finalizer(Instance* key, Object* lambda, bool weak_map) {
   // We should already have checked whether the object is already registered.
-  ASSERT(!has_finalizer(key, lambda));
-  auto node = _new ToitFinalizerNode(key, lambda);
+  CallableFinalizerNode* node;
+  if (weak_map) {
+    ASSERT(!key->can_be_toit_finalized(program()));
+    ASSERT(!key->has_active_finalizer());
+    node = _new WeakMapFinalizerNode(key, lambda, this);
+  } else {
+    ASSERT(key->can_be_toit_finalized(program()));
+    ASSERT(!key->has_active_finalizer());
+    node = _new ToitFinalizerNode(key, lambda, this);
+  }
   if (node == null) return false;  // Allocation failed.
-  registered_toit_finalizers_.append(node);
+  // Add it at the head of the list in case there is an old finalizer lower
+  // down on the list.
```

Comment:
```suggestion
  // Add it at the head of the list in case there is an old "removed" finalizer lower
  // down on the list, waiting to be unlinked at the next GC.
```